src/main/java/mesosphere/dcos/client/DCOSClient.java

```diff
@@ -52,7 +67,15 @@ public Exception decode(String methodKey, Response response) {
             } catch (NullPointerException | IOException e) {
                 details = "Unable to read response body";
             }
-            return new DCOSException(response.status(), response.reason(), methodKey, details);
+
+            DCOSException e = new DCOSException(response.status(), response.reason(), methodKey, details);
+
+            if (authTokenHeaderInterceptor != null && response.status() == 401 && authTokenHeaderInterceptor.hasToken()) {
+                authTokenHeaderInterceptor.clearToken();
```

Comment:
Thinking through any possible thread safety issues here if the same DCOSClient is used on multiple threads.  

* If we get a 401 but another thread clears the token before we check `hasToken` we'll return a DCOSException instead of a RetryableException.  
* We may  clear a token that's being used by another thread.  If it was expired then it would have been unable to use that token anyway.  The only concern would be that having it suddenly become null might lead to an exception that isn't handled appropriately.   (like a nullpointerexception from `dcosAuthToken.requiresRefresh()`

Not sure either of those are terrible though.  It'd be kind of a weird thing to try to handle too.  Would have to start a critical section here and block DCOSAuthTokenHeaderInterceptor from using the token until it had been replaced (which of course requires DCOSAuthTokenHeaderInterceptor to check the token).  Thoughts?